File: go/backpressure.go
// SPDX-FileCopyrightText: Copyright (C) 2026 provide.io llc
// SPDX-License-Identifier: Apache-2.0

package telemetry

import "sync"

const _defaultQueueSize = 0

// QueuePolicy defines per-signal maximum in-flight queue sizes.
type QueuePolicy struct {
	LogsMaxSize    int
	TracesMaxSize  int
	MetricsMaxSize int
}

var (
	_queueMu      sync.RWMutex
	_queuePolicy  QueuePolicy
	_logsQueue    chan struct{}
	_tracesQueue  chan struct{}
	_metricsQueue chan struct{}
)

func init() {
	_resetQueuePolicy()
}

// QueueTicket represents one acquired backpressure slot.
// Release the returned ticket to free the exact queue slot that was acquired.
type QueueTicket struct {
	signal string
	ch     chan struct{}
}

// _buildQueue returns a buffered channel of the given size, or nil for unlimited (size <= 0).
func _buildQueue(size int) chan struct{} {
	if size <= 0 {
		return nil
	}
	return make(chan struct{}, size)
}

// _resetQueuePolicy rebuilds all queues with default sizes.
// Must be called with no lock held.
func _resetQueuePolicy() {
	_queueMu.Lock()
	defer _queueMu.Unlock()
	_drainAndRebuild(
		QueuePolicy{
			LogsMaxSize:    _defaultQueueSize,
			TracesMaxSize:  _defaultQueueSize,
			MetricsMaxSize: _defaultQueueSize,
		},
	)
}

// _drainAndRebuild replaces all queues with new channels based on policy.
// Old channels are abandoned (not closed) to prevent a panic: closing a channel
// while a concurrent sender holds a stale pointer to it would panic. GC reclaims
// the old channel once all in-flight references drop.
// Caller must hold _queueMu write lock.
func _drainAndRebuild(policy QueuePolicy) {
	// Intentionally do not close old channels — let them be GC'd once all
	// in-flight references drop.
	_queuePolicy = policy
	_logsQueue = _buildQueue(policy.LogsMaxSize)
	_tracesQueue = _buildQueue(policy.TracesMaxSize)
	_metricsQueue = _buildQueue(policy.MetricsMaxSize)
}

// SetQueuePolicy replaces the semaphore channels with new sizes.
func SetQueuePolicy(policy QueuePolicy) {
	_queueMu.Lock()
	defer _queueMu.Unlock()
	_drainAndRebuild(policy)
}

// GetQueuePolicy returns the current policy under read lock.
func GetQueuePolicy() QueuePolicy {
	_queueMu.RLock()
	defer _queueMu.RUnlock()
	return _queuePolicy
}

// _queueForSignal returns the configured max size and current channel for a signal.
// Caller must hold at least a read lock.
func _queueForSignal(signal string) (int, chan struct{}, bool) {
	switch signal {
	case signalLogs:
		return _queuePolicy.LogsMaxSize, _logsQueue, true
	case signalTraces:
		return _queuePolicy.TracesMaxSize, _tracesQueue, true
	case signalMetrics:
		return _queuePolicy.MetricsMaxSize, _metricsQueue, true
	default:
		return 0, nil, false
	}
}

// TryAcquire attempts a non-blocking acquire on the signal's semaphore.
// Returns a ticket if a slot was acquired, or nil if the queue is at capacity
// or the signal is unknown. A maxSize of 0 (or negative) means unlimited:
// acquisition succeeds with a ticket that does not hold a channel slot.
// The RLock is held through the entire read + send to prevent a TOCTOU race where
// SetQueuePolicy swaps the channel pointer between the read and the send.
func TryAcquire(signal string) *QueueTicket {
	_queueMu.RLock()
	defer _queueMu.RUnlock()

	maxSize, ch, ok := _queueForSignal(signal)
	if !ok {
		return nil
	}
	if maxSize <= 0 {
		return &QueueTicket{signal: signal}
	}

	select {
	case ch <- struct{}{}:
		return &QueueTicket{signal: signal, ch: ch}
	default:
		_incDropped(signal)
		return nil
	}
}

// Release frees the exact queue slot represented by ticket.
// It is a no-op for nil tickets and unlimited queues.
func Release(ticket *QueueTicket) {
	if ticket == nil || ticket.ch == nil {
		return
	}

	select {
	case <-ticket.ch:
	default:
	}
}

// _incDropped increments the appropriate "dropped" counter.
func _incDropped(signal string) {
	switch signal {
	case signalLogs:
		_incLogsDropped()
	case signalTraces:
		_incSpansDropped()
	case signalMetrics:
		_incMetricsDropped()
	}
}

File: csharp/src/Provide.Telemetry/Backpressure.cs
// SPDX-FileCopyrightText: Copyright (C) 2026 provide.io llc
// SPDX-License-Identifier: Apache-2.0
// SPDX-Comment: Part of provide-telemetry.


namespace Provide.Telemetry;

public static class Backpressure
{
    private static readonly object Gate = new();
    private static QueuePolicy _policy = new();
    private static int _logsCount;
    private static int _tracesCount;
    private static int _metricsCount;

    public static void SetQueuePolicy(QueuePolicy policy)
    {
        lock (Gate)
        {
            _policy = new QueuePolicy
            {
                LogsMaxSize = Math.Max(0, policy.LogsMaxSize),
                TracesMaxSize = Math.Max(0, policy.TracesMaxSize),
                MetricsMaxSize = Math.Max(0, policy.MetricsMaxSize),
            };
            _logsCount = 0;
            _tracesCount = 0;
            _metricsCount = 0;
        }
    }

    public static QueuePolicy GetQueuePolicy()
    {
        lock (Gate)
        {
            return new QueuePolicy
            {
                LogsMaxSize = _policy.LogsMaxSize,
                TracesMaxSize = _policy.TracesMaxSize,
                MetricsMaxSize = _policy.MetricsMaxSize,
            };
        }
    }

    /// <summary>The configured bound for a signal; 0 means unlimited.</summary>
    public static int MaxSize(string signal)
    {
        lock (Gate)
        {
            return signal switch
            {
                Signals.Logs => _policy.LogsMaxSize,
                Signals.Traces => _policy.TracesMaxSize,
                Signals.Metrics => _policy.MetricsMaxSize,
                _ => 0,
            };
        }
    }

    /// <summary>
    /// Take a queue slot, or null when the bounded queue is full.
    /// </summary>
    /// <remarks>
    /// A refusal is not counted here. <see cref="SignalPipeline"/> owns the
    /// health accounting for every admission decision so that consent, sampling
    /// and backpressure rejections are recorded at one stage, in one order,
    /// exactly once — counting in both places double-counted every full queue.
    /// </remarks>
    public static QueueTicket? TryAcquire(string signal)
    {
        lock (Gate)
        {
            var (max, count) = signal switch
            {
                Signals.Logs => (_policy.LogsMaxSize, _logsCount),
                Signals.Traces => (_policy.TracesMaxSize, _tracesCount),
                Signals.Metrics => (_policy.MetricsMaxSize, _metricsCount),
                _ => (-1, 0),
            };
            if (max < 0) return null;
            // 0 means unlimited
            if (max == 0 || count < max)
            {
                switch (signal)
                {
                    case Signals.Logs: _logsCount++; break;
                    case Signals.Traces: _tracesCount++; break;
                    case Signals.Metrics: _metricsCount++; break;
                }
                return new QueueTicket(signal);
            }
            return null;
        }
    }

    public static void Release(QueueTicket? ticket)
    {
        if (ticket is null || ticket.Released) return;
        lock (Gate)
        {
            if (ticket.Released) return;
            ticket.Released = true;
            switch (ticket.Signal)
            {
                case Signals.Logs when _logsCount > 0: _logsCount--; break;
                case Signals.Traces when _tracesCount > 0: _tracesCount--; break;
                case Signals.Metrics when _metricsCount > 0: _metricsCount--; break;
            }
        }
    }

    internal static void Reset()
    {
        lock (Gate)
        {
            _policy = new QueuePolicy();
            _logsCount = _tracesCount = _metricsCount = 0;
        }
    }
}

public sealed class QueueTicket
{
    internal QueueTicket(string signal) { Signal = signal; }
    public string Signal { get; }
    internal bool Released { get; set; }
}
